### PR TITLE
store/store_download.go: use system snap provided xdelta3 priority + fallback

### PR DESCRIPTION
**Pull request by anonymouse64**
Prefer using xdelta3 from the system snap if it is available and we are able to
run the config subcommand successfully. This ensures that we try to use a known
good xdelta3 command which should always work with what the store sends down,
but does still offer the fallback of using xdelta3 from the host system if for
some reason the xdelta3 command from the system snap becomes broken at some
point. In that case, attempt using the xdelta3 command from the host system,
performing the same config check before assuming that it is usable.

This has the effect that on older systems like Centos 7 and Amazon Linux 2,
where for example xdelta3 from the distro does not support LZMA, when a delta
generated using LZMA is encountered, we will first try the system snap provided
xdelta3 which should succeed.

We also cache the result of inspecting the available xdelta3 commands, to
prevent unnecessary commands.

Finally, log more messages, namely that if applying the delta fails, log the
error condition as a message so someone with log access can provide information
as well as log if we find an unsuitable xdelta3 somewhere (either from the
system snap, or from the host system). All of these should aid in debugging
future delta issues if they happen to crop up.

Marking as complex because retries are hard and the store download code is
also hard.

### CODE REVIEW

**Comment by anonymouse64** [comment]
The refresh-delta tests for Centos 7 and Amazon Linux 2 here are all happy, so I will work on adding the appropriate unit tests to mark this as unblocked

**Comment by stolowski** [comment]
Could we use `xdelta3 config` (or xdelta -V for version check) to determine the version of xdelta3 to use once on startup instead of re-trying them on every download?

This is what xdelta3 config shows on focal:
```
xdelta3 config
Xdelta version 3.0.11, Copyright (C) 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015 Joshua MacDonald
Xdelta comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it
under certain conditions; see "COPYING" for details.
EXTERNAL_COMPRESSION=1
REGRESSION_TEST=1
SECONDARY_DJW=1
SECONDARY_FGK=1
SECONDARY_LZMA=1
UNALIGNED_OK=1
....
```

**Comment by pedronis** [comment]
I agree with Pawel, not sure we actually need to check the version, but we probably should turn useDeltas into a method of the Store and there try to run "xdelta3 config" from the snaps, if that works we are fine, otherwise try the system one and possibly check for values?  if both fails just remember to return false on future invocations (we might need a lock around this though or to call useDeltas or the checking code alone from the store creation).

File: store/store_download.go [hunk 1]
```diff
@@ -646,16 +653,38 @@ var applyDelta = func(name string, deltaPath string, deltaInfo *snap.DeltaInfo,
...
 	if err != nil {
 		return err
 	}
 
-	if err := cmd.Run(); err != nil {
-		if err := os.Remove(partialTargetPath); err != nil {
-			logger.Noticef("failed to remove partial delta target %q: %s", partialTargetPath, err)
+	// try to run the xdelta command, optionally falling back and cleaning up
+	// afterwards
+	if runErr := cmd.Run(); runErr != nil {
```
Comment (mvo5) [inline store/store_download.go:663]: Maybe we can use a slightly more linear way of doing this? Not sure if that is possible but I sketched an idea below.

 Maybe something similar to:
```go
diff --git a/store/store_download.go b/store/store_download.go
index c7095c8439..6b85cae9e5 100644
--- a/store/store_download.go
+++ b/store/store_download.go
@@ -83,8 +83,11 @@ func init() {
diff --git a/store/store_download.go b/store/store_download.go
index c7095c8439..ae4f189d10 100644
--- a/store/store_download.go
+++ b/store/store_download.go
@@ -83,7 +83,7 @@ func init() {
 func useDeltas() bool {
        // only xdelta3 is supported for now, so check the binary exists here
        // TODO: have a per-format checker instead
-       if _, _, err := getXdelta3Cmd(); err != nil {
+       if _, err := snapdtool.CommandFromSystemSnap("/usr/bin/xdelta3"); err != nil && !osutil.ExecutableExists("xdelta3") {
                return false
        }
 
@@ -598,7 +601,7 @@ func doDownloadReqImpl(ctx context.Context, storeURL *url.URL, cdnHeader string,
 }
 
 // downloadDelta downloads the delta for the preferred format, returning the path.
-func (s *Store) downloadDelta(deltaName string, downloadInfo *snap.DownloadInfo, w io.ReadWriteSeeker, pbar progress.Meter, user *auth.UserState, dlOpts *DownloadOptions) error {
+func (s *Store) downloadDelta(deltaName string, downloadInfo *snap.DownloadInfo, w io.ReadWriteSeeker, pbar progress.Meter, user *auth.UserState, dlOpts *DownloadOptions) (err error) {
 
        if len(downloadInfo.Deltas) != 1 {
                return errors.New("store returned more than one download delta")
@@ -623,22 +626,8 @@ func (s *Store) downloadDelta(deltaName string, downloadInfo *snap.DownloadInfo,
        return download(context.TODO(), deltaName, deltaInfo.Sha3_384, url, user, s, w, 0, pbar, dlOpts)
 }
 
-func getXdelta3Cmd(args ...string) (*exec.Cmd, bool, error) {
-       cmd, err := snapdtool.CommandFromSystemSnap("/usr/bin/xdelta3", args...)
-       if err != nil {
-               // try checking the system provided version
-               if osutil.ExecutableExists("xdelta3") {
-                       return exec.Command("xdelta3", args...), false, nil
-               }
-
-               return nil, false, fmt.Errorf("no xdelta3 available")
-       }
-
-       return cmd, true, nil
-}
-
 // applyDelta generates a target snap from a previously downloaded snap and a downloaded delta.
-var applyDelta = func(name string, deltaPath string, deltaInfo *snap.DeltaInfo, targetPath string, targetSha3_384 string) error {
+var applyDelta = func(name string, deltaPath string, deltaInfo *snap.DeltaInfo, targetPath string, targetSha3_384 string) (err error) {
        snapBase := fmt.Sprintf("%s_%d.snap", name, deltaInfo.FromRevision)
        snapPath := filepath.Join(dirs.SnapBlobDir, snapBase)
 
@@ -651,39 +640,26 @@ var applyDelta = func(name string, deltaPath string, deltaInfo *snap.DeltaInfo,
        }
 
        partialTargetPath := targetPath + ".partial"
-
-       xdelta3Args := []string{"-d", "-s", snapPath, deltaPath, partialTargetPath}
-       cmd, isSystemSnapProvided, err := getXdelta3Cmd(xdelta3Args...)
-       if err != nil {
-               return err
-       }
-
-       // try to run the xdelta command, optionally falling back and cleaning up
-       // afterwards
-       if runErr := cmd.Run(); runErr != nil {
-               if isSystemSnapProvided {
-                       // fallback to trying the version on $PATH instead
-                       if !osutil.ExecutableExists("xdelta3") {
-                               // internal error, we should have already checked if deltas are
-                               // available before calling this function
-                               return fmt.Errorf("internal: no xdelta3 available")
-                       }
-                       pathCmd := exec.Command("xdelta3", xdelta3Args...)
-                       if runErr := pathCmd.Run(); runErr != nil {
-                               if err := os.Remove(partialTargetPath); err != nil {
-                                       logger.Noticef("after encountering error applying delta: %v, also failed to remove partial delta target %q: %s", runErr, partialTargetPath, err)
-                               }
-                       }
-                       // otherwise we succeeded with the $PATH version, but still failed
-                       // with the system provided one, so log that error in case we end up
-                       // having issues there
-                       logger.Noticef("overcame error from %q from system snap (%v) by using $PATH version (%q)", cmd.Path, runErr, pathCmd.Path)
-               } else {
-                       // we already used the $PATH, so no fallback
+       defer func() {
+               if err != nil {
                        if err := os.Remove(partialTargetPath); err != nil {
-                               logger.Noticef("after encountering error applying delta: %v, also failed to remove partial delta target %q: %s", runErr, partialTargetPath, err)
+                               logger.Noticef("cannot remove partial file: %v", err)
                        }
-                       return err
+               }
+       }()
+
+       // try xdelta from the snap first
+       args := []string{"-d", "-s", snapPath, deltaPath, partialTargetPath}
+       cmd, err := snapdtool.CommandFromSystemSnap("/usr/bin/xdelta3", args...)
+       if err == nil {
+               err = cmd.Run()
+       }
+       if err != nil {
+               // try system xdelta next
+               logger.Noticef("cannot use internal xdelta: %v", err)
+               cmd = exec.Command("xdelta3", args...)
+               if err = cmd.Run(); err != nil {
+                       return fmt.Errorf("cannot use system xdelta: %v", err)
                }
        }
```
```

**Review by mvo5 — COMMENTED** [review]
Thanks for working on this! One small suggestion inline how this could be maybe a bit more linear. But feel free to adjust/ignore, maybe it's just me :)

File: store/store_download.go (hunk 1 above)
Comment (anonymouse64) [reply store/store_download.go]: thanks for the suggestion, but as per Pawel and Samuele's comment I had to go with a entirely different implementation

**Comment by anonymouse64** [comment]
This is ready for review again, took a bit of finagling with the mock command, turns out that testutil.MockCommand is unhappy with an empty $PATH because it uses `basename` and can't find that command if $PATH is empty.

File: store/download_test.go [hunk 2]
```diff
@@ -435,20 +441,102 @@ func (s *downloadSuite) TestUseDeltas(c *C) {
...
+			// as well as the actual interpreter for executing the basic
+			// "xdelta3 config" command
+			// for the interpreter, since that's how we execute xdelta3, mock
+			// that as a command, but for the xdelta3 command we need the
+			// PT_INTERP symbol set for the binary for elf handling to work,
+			// which won't be the case on a basic shell script
+
+			// TODO: what if the unit tests are run on a system where /bin/true
+			// does not have a PT_INTERP which equals lib64/ld-linux-x86-64.so.2
+			// i.e. lp builders ?
```
Comment (anonymouse64) [inline store/download_test.go:464]: this will presumably be a real problem that needs to be solved before landing this PR, would be happy to hear suggestions on how to handle this gracefully

File: store/store_download.go [hunk 3]
```diff
@@ -80,14 +80,63 @@ func init() {
...
+
+	// check if the xdelta3 config command works from the system snap
+	cmd, err := snapdtool.CommandFromSystemSnap("/usr/bin/xdelta3", "config")
+	if err == nil {
+		// we have a xdelta3 from the system snap, make sure it works
+		if runErr := cmd.Run(); runErr == nil {
+			// success using the system snap provided one
+			// note that the first three arguments to the return value from
+			// CommandFromSystemSnap are the interpreter for that system snap
+			s.xdelta3CmdLocation = cmd.Args[3]
```
Comment (pedronis) [inline store/store_download.go:115]: shouldn't we record something to invoke it using or as if using snapdtool.CommandFromSystemSnap ?

**Review by pedronis — COMMENTED** [review]
question

File: store/download_test.go (hunk 2 above)
Comment (mardy) [reply store/download_test.go:464]: I think that the unit tests should not call the real `CommandFromSystemSnap()` function, but a mocked version. I see that `overlord/snapstate/backend/fontconfig.go` and `snap/squashfs/squashfs.go` use this function with a level of indirection that allows injecting a mocked version; maybe you should try the same here?

File: store/store_download.go [hunk 4]
```diff
@@ -646,16 +692,20 @@ var applyDelta = func(name string, deltaPath string, deltaInfo *snap.DeltaInfo,
...
+
+	// sanity check that deltas are available and that the path for the xdelta3
+	// command is set
+	if ok := s.useDeltas(); !ok {
+		return fmt.Errorf("internal error: applyDelta used when deltas are not available")
 	}
 
-	if err := cmd.Run(); err != nil {
+	// run the xdelta3 command, cleaning up if we fail and logging about it
+	if runErr := exec.Command(s.xdelta3CmdLocation, xdelta3Args...).Run(); runErr != nil {
```
Comment (mvo5) [inline store/store_download.go:703]: This may be too simple for the case when we use the xdelta3 from the core/snapd snap. The CommandFromSystemSnap() is actually running it through `/lib64/ld-linux-x86-64.so.2` and adding a --library-path that will point to the core/snapd snap.

File: store/download_test.go (hunk 2 above)
Comment (pedronis) [reply store/download_test.go:464]: this might be easier indeed

File: store/store.go [hunk 5]
```diff
@@ -158,6 +158,12 @@ type Store struct {
 	proxyConnectHeader http.Header
 
 	userAgent string
+
+	xdeltaCheckLock sync.Mutex
+	// whether we should use deltas or not
+	shouldUseDeltas *bool
+	// which xdelta3 we picked when we checked the deltas
+	xdelta3CmdLocation string
```
Comment (pedronis) [inline store/store.go:166]: it sounds like this one should become a callback that can be set to return either the a exec.Cmd for the host or for system snap xdelta3:

xdelta3Cmd  func()  (*exec.Cmd, error)

**Review by pedronis — COMMENTED** [review]
couple of considerations

File: store/store_download.go [hunk 6]
```diff
@@ -80,14 +80,63 @@ func init() {
...
+	}
+
+	// TODO: have a per-format checker instead, we currently only support
+	// xdelta3 as a format for deltas
+
+	// check if the xdelta3 config command works from the system snap
+	cmd, err := snapdtool.CommandFromSystemSnap("/usr/bin/xdelta3", "config")
+	if err == nil {
+		// we have a xdelta3 from the system snap, make sure it works
+		if runErr := cmd.Run(); runErr == nil {
```
Comment (stolowski) [inline store/store_download.go:111]: That's fine but I wonder if we should be paranoid over tools we ship in system snap. They should just work, no (if not, we probably have a bigger problem?).

Comment (anonymouse64) [reply store/store_download.go:111]: well the fact that we cache the result of this means that there isn't a performance penalty to this paranoia so I think we should stick with more paranoia, especially given the fact that it sort of changed under our hood and that how we run things from the system snaps using the interpreter manually is a bit weird and could be fragile long term

Comment (stolowski) [reply store/store_download.go:111]: Yeah, you have a point there, fair enough.

File: store/download_test.go [hunk 7]
```diff
@@ -435,20 +441,136 @@ func (s *downloadSuite) TestUseDeltas(c *C) {
...
+			// handling that is per-arch, etc. of the real CommandFromSystemSnap
+			// implementation.
+
+			coreInterpCmd = testutil.MockCommand(c, interpInCorePath, "")
+
+			r := store.MockSnapdtoolCommandFromSystemSnap(func(name string, args ...string) (*exec.Cmd, error) {
+				c.Assert(name, Equals, "/usr/bin/xdelta3")
+				c.Assert(args, DeepEquals, []string{"config"})
+
+				interpArgs := append([]string{"--library-path", "/some/dir/from/etc/ld.so", exeInCorePath}, args...)
```
Comment (pedronis) [inline store/download_test.go:473]: we can probably reduce the realism given we are mocking things anyway, just have some marker

File: store/download_test.go [hunk 8]
```diff
@@ -435,20 +441,136 @@ func (s *downloadSuite) TestUseDeltas(c *C) {
...
+			// "xdelta3 config" command.
+			// For the interpreter, since that's how we execute xdelta3, mock
+			// that as a command, but we don't need to mock the xdelta3 command
+			// in the core snap since that doesn't get executed by our fake
+			// interpreter. Mocking the interpreter and executing that as a
+			// MockCommand has the advantage that it avoids the specific ELF
+			// handling that is per-arch, etc. of the real CommandFromSystemSnap
+			// implementation.
+
+			coreInterpCmd = testutil.MockCommand(c, interpInCorePath, "")
```
Comment (pedronis) [inline store/download_test.go:467]: don't we need to call Restore on this?

File: store/download_test.go [hunk 9]
```diff
@@ -435,20 +441,136 @@ func (s *downloadSuite) TestUseDeltas(c *C) {
...
+				coreInterpCmd.ForgetCalls()
+			})
+		}
+
 		if scenario.exeInHost {
-			os.Setenv("PATH", origPath)
-		} else {
-			os.Setenv("PATH", altPath)
+			// just mock the xdelta3 command directly
+			hostXdelta3Cmd = testutil.MockCommand(c, "xdelta3", "")
```
Comment (pedronis) [inline store/download_test.go:493]: and this, to cleanup ?

File: store/download_test.go [hunk 10]
```diff
@@ -435,20 +441,136 @@ func (s *downloadSuite) TestUseDeltas(c *C) {
...
+				// just called xdelta3 directly
+				c.Check(hostXdelta3Cmd.Calls(), DeepEquals, [][]string{
+					{"xdelta3", "config"},
+				}, comment)
+
+				// and args are passed to the command cached too
+				expArgs := []string{hostXdelta3Cmd.Exe(), "foo", "bar"}
+				c.Check(sto.Xdelta3Cmd("foo", "bar").Args, DeepEquals, expArgs, comment)
+			}
+		}
```
Comment (pedronis) [inline store/download_test.go:568]: it feels like we should sanity check in the else here,  that both exe are false or env is false?

**Review by pedronis — COMMENTED** [review]
some comments, the scenario test is a bit hard to follow :/ oh well

File: store/download_test.go (hunk 9 above)
Comment (anonymouse64) [reply store/download_test.go:493]: we do call Restore() on this one, but we call it before the second UseDeltas() invocation, in order to test the caching properly, hence it is not in a cleanup

File: store/download_test.go (hunk 7 above)
Comment (anonymouse64) [reply store/download_test.go:473]: personally I think it is preferable to leave it in, there isn't a great reduction in complexity of the test by passing some other arguments here, and using what matches reality here is a better approximation even if things are being mocked so I feel creates a better test. I will add a comment about where these sorts of things come from though

File: store/download_test.go [hunk 11]
```diff
@@ -385,17 +387,21 @@ func (s *downloadSuite) TestActualDownloadServerNoResumeHandeled(c *C) {
...
 	restore := release.MockOnClassic(false)
 	defer restore()
-	altPath := c.MkDir()
+
 	origSnapMountDir := dirs.SnapMountDir
 	defer func() { dirs.SnapMountDir = origSnapMountDir }()
 	dirs.SnapMountDir = c.MkDir()
 	exeInCorePath := filepath.Join(dirs.SnapMountDir, "/core/current/usr/bin/xdelta3")
+	interpInCorePath := filepath.Join(dirs.SnapMountDir, "/core/current/lib64/ld-linux-x86-64.so.2")
+	os.MkdirAll(filepath.Dir(interpInCorePath), 0755)
```
[the remark this answers is not in the document]
Comment (anonymouse64) [reply store/download_test.go]: fixed

File: store/download_test.go (hunk 8 above)
Comment (anonymouse64) [reply store/download_test.go:465]: I added a call to Restore() anyways

File: store/download_test.go (hunk 10 above)
Comment (anonymouse64) [reply store/download_test.go:589]: added a check to this

File: store/download_test.go (hunk 9 above)
Comment (anonymouse64) [reply store/download_test.go:503]: I added a comment to explain where the Restore() is located for the casual reader

**Review by mardy — APPROVED** [review]
LGTM!

File: store/store_download.go [hunk 12]
```diff
@@ -80,14 +82,79 @@ func init() {
...
+			// but drop the last arg which we know is "config"
+			exe := cmd.Path
+			args := cmd.Args[:len(cmd.Args)-1]
+			env := cmd.Env
+			dir := cmd.Dir
+			s.xdelta3CmdFunc = func(xDelta3args ...string) *exec.Cmd {
+				return &exec.Cmd{
+					Path: exe,
+					Args: append(args, xDelta3args...),
+					Env:  env,
```
Comment (pedronis) [inline store/store_download.go:128]: do we need to copy Env and Dir? afaiu the snapdtool helper itself uses exec.Command which doesn't set those

**Review by pedronis — APPROVED** [review]
thanks, maybe we can simplify the exec.Cmd copying a bit more?

File: store/store_download.go (hunk 12 above)
Comment (anonymouse64) [reply store/store_download.go:128]: No, but I thought it good practice since snapdtool.CommandFromSystemSnap could some day return more things in the Cmd and we may not catch that here when we make that change. I can remove this if you want since snapdtool does not currently set those as you pointed out

**Comment by anonymouse64** [comment]
The failures here are unrelated, some of which are fixed by https://github.com/snapcore/snapd/pull/10532, so I think this should be merged to at least get the delta-refresh tests to stop failing on AL2 and Centos7
